AWSPinpoint/AWSPinpointNotificationManager.m

```diff
@@ -224,29 +228,22 @@ - (void)recordMessageOpenedEventForNotification:(NSDictionary *) userInfo
 
 #pragma mark - Helpers
 - (void)addEventSourceMetadataForEvent:(AWSPinpointEvent *) event
-                      withNotification:(NSDictionary *) userInfo {
-    if (![AWSPinpointNotificationManager validPinpointPushForNotification:userInfo]) {
-        return;
-    }
-    NSDictionary *campaign = userInfo[AWSDataKey][AWSPinpointKey][AWSPinpointCampaignKey];
-    AWSDDLogVerbose(@"Adding campaign attributes to event[%@]: %@",event.eventType, campaign);
-    
-    for (NSString *key in [campaign allKeys]) {
-        [event addAttribute:campaign[key] forKey:key];
+                          withMetadata:(NSDictionary *) metadata {
+    for (NSString *key in [metadata allKeys]) {
+        [event addAttribute:metadata[key] forKey:key];
     }
 }
 
-- (void)addGlobalEventSourceMetadataForNotification:(NSDictionary *) userInfo {
-    if (![AWSPinpointNotificationManager validPinpointPushForNotification:userInfo]) {
-        return;
-    }
-    
-    NSDictionary *campaign = userInfo[AWSDataKey][AWSPinpointKey][AWSPinpointCampaignKey];
-    AWSDDLogVerbose(@"Adding campaign global attributes: %@", campaign);
-    [self.context.analyticsClient setEventSourceAttributes:campaign];
-    
-    for (NSString *key in [campaign allKeys]) {
-        [self.context.analyticsClient addGlobalAttribute:campaign[key] forKey:key];
+- (void)addGlobalEventSourceMetadata:(NSDictionary *) metadata {
+    if (metadata.count) {
+        // Remove previous global event source attributes from _globalAttributes
+        // This is to prevent _globalAttributes containing attributes from multiple event sources (campaign/journey)
+        [self.context.analyticsClient removeAllGlobalEventSourceAttributes];
+        [self.context.analyticsClient setEventSourceAttributes:metadata];
+
+        for (NSString *key in [metadata allKeys]) {
+            [self.context.analyticsClient addGlobalAttribute:metadata[key] forKey:key];
+        }
```

Comment:
Correct me if i'm wrong, but there seems to be a threading issue where we could have an incoming call from both:
`interceptDidFinishLaunchingWithOptions` and `interceptDidReceiveRemoteNotification:fetchCompletionHandler:`, we could call this function at the same time causing some inconsistent behavior?

Also, I'm guessing it's relatively safe to call ` removeAllGlobalEventSourceAttributes` from this context (?) -- not being super familiar with this code, are there any flaws that you can see by calling this function here?

Reply:
If the user receives a push in the background, and taps on it to open. `addGlobalEventSourceMetadata` will be called twice, but the attributes will be the same, so it will not cause any consistency. `AWSPinpointAnalyticsClient` does synchronize on global attributes updates. The downside of adding this call is we are removing previous values every time we update, instead of overwriting existing ones previously.